### DISCUSSION TITLE
[Components] Progress API

### DISCUSSION THREAD

**Discussion by soykje** [adevinta/spark#992] [Decision making]
Hey! :wave:

Waiting for Design specs, we want to open the discussions about possible API for the Progress component. Here are some resources:
- https://react.carbondesignsystem.com/?path=/docs/components-progressbar--default
- https://www.radix-ui.com/docs/primitives/components/progress
- https://www.decathlon.design/726f8c765/v/0/p/2388f2-progressbar/b/199c66
- https://react-spectrum.adobe.com/react-aria/ProgressBar.html
- https://atlassian.design/components/progress-bar/examples

From this starting point we could imagine the following API:
```tsx
/*
* Short description of the progress we want to display (ex: "Loading...")
*/
label?: ReactNode
/*
* The relative progress value (within min|max range). Default is 0.
*/
value: number
/*
* Displayed readable value (ex: "42%")
*/
valueLabel?: ReactNode
/*
* Min allowed value (default 0)
*/
min?: number
/*
* Max allowed value (default 0)
*/
max?: number
/*
* Progress bar|gauge thickness
*/
size?: 'sm' | 'md' (default 'md')
/*
* Progress displayed as bar|gauge (default 'linear')
*/
variant?: 'linear' | 'circular'
/*
* Whether the progress is unknown or cannot be determined (default false)
*/
isIndeterminate?: boolean
```

As you'll notice there is a strong inspiration from React Spectrum. I just simplified a bit the `valueLabel` management, and added the `variant` possibility that could be useful on LBC/Polaris existing codebase. This variant could also, at minimal cost, avoid us to expose a `progress-circular` evil twin :smiling_imp: 

Wdyt? Let's discuss!

**Comment by acd02** [comment]
Should we also consider using a `render prop` approach, like React Spectrum does? This approach would provide great flexibility to consumers of the component. However, it could also be argued that it may provide too much flexibility, which may not be desired 🤷

**Reply by soykje** [reply]
Yep, this is also what I thought: a nice idea, but that may bring too much complexity in the first iterations of the component. Still, it can be discussed and maybe implemented in a near future, if needed?

**Reply by acd02** [reply]
Yes, we could consider doing that later. That was just a thought 👍

**Comment by Powerplex** [comment]
Imho the shape difference betwee the `linear` and `circular` justifies having two separate components: `ProgressBar` and `ProgressCircular` (or something like that). I guess markup will be different + it does not take the same ratio on the screen at all. But my opinion is we can start without the circular implementation for a v1.

**Reply by acd02** [reply]
Maybe we could do something like:
```tsx
import { Progress } from 'components/progress'

Progress.Bar
Progress.Circular
```

**Reply by soykje** [reply]
Yes we could end with a compound one, but we also may simply use the `variant` prop. Sure the `line` would become a `circle` within the `<svg />` but as we expect same functionnalities it may be a good choice to stick with only one component, imho. :slightly_smiling_face:

**Reply by acd02** [reply]
Hm, I’ll side with Julien here. I prefer having two separate components (whether or not we opt for the Compound pattern is another question) if their shape/markup is significantly different, rather than using an enum

**Reply by soykje** [reply]
Alright then! Wdyt of a compound pattern then, as you suggested? :slightly_smiling_face:

**Reply by acd02** [reply]
I'm down with that 👍

**Reply by soykje** [reply]
It seems we misunderstood on this one :sweat_smile: So let's continue the discussion!

**Reply by acd02** [reply]
To clarify, when I said **Compound**, I didn't mean that both components should share a context or anything like that. It was rather a cosmetic point. I just meant that we could export an object containing both components, i.e.:
```tsx
import { Bar } from 'components/Bar'
import { Circle } from 'components/Circle'

export const Progress = {
  Bar,
  Circle
}
```

**Comment by soykje** [comment]
Ok. To unblock this topic, I suggest we start with a Spectrum-inspired `ProgressBar` component, wdyt?

**Reply by acd02** [reply]
So, are we only starting with a progress bar component for now?

**Reply by soykje** [reply]
This is what I suggest, but I guess it would be great to get some feedbacks from @turolopezsanabria and Jeremias, before getting started...

**Comment by soykje** [comment]
**tldr;** We have to decide if we want to ship both progress "bar" and progress "circle" (there is a nice example about that on Decathlon DS).
If the "circle" variant is not needed, then no question anymore, I guess :slightly_smiling_face: And if it is, then we could discuss about having them shipped as a single package, or separate ones.

**Reply by soykje** [reply]
Personnally I think we should consider it, at least in our initial conception (even if the feature itself isn't shipped immediately). See https://react-spectrum.adobe.com/react-aria/ProgressBar.html#circular and https://www.decathlon.design/726f8c765/v/0/p/2388f2-progressbar/b/199c66/t/678ca1 :pray:

**Reply by acd02** [reply]
The React Spectrum example is not the same as your proposition though. React Spectrum employs a headless component that exposes a `children prop`, and from that, they build two distinct components: a **ProgressBar** and a **ProgressCircular**

**Reply by soykje** [reply]
Yes sorry for the misunderstanding... My point here was just to help in decision about having a unique component with circular and linear "variants", or not. The way we'll compose this could be discussed after that :slightly_smiling_face:
